drivers/cache/cache_bflb_l1c.c

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright (c) 2025 MASSDRIVER EI (massdriver.space)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/cache.h>
+#include <zephyr/logging/log.h>
+
+#include <bflb_soc.h>
+#include <hbn_reg.h>
+#include <l1c_reg.h>
+#include <zephyr/drivers/clock_control/clock_control_bflb_common.h>
+
+#define DT_DRV_COMPAT bflb_l1c
+
+#define INVLD_TIMEOUT 4194304
+
+LOG_MODULE_REGISTER(cache_bflb_l1c, CONFIG_CACHE_LOG_LEVEL);
+
+struct cache_config {
+	uint32_t ways_mask;
+};
+
+static struct cache_config cache_cfg = {
+	.ways_mask = (1 << DT_INST_PROP(0, ways_disabled)) - 1,
+};
+
+/* L1C supports basically a single type of invalidate & flush */
+int bflb_cache_invalidate(uint32_t timeout)
+{
+	uint32_t tmp;
+
+	tmp = *((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET));
+	tmp |= 1 << L1C_BYPASS_POS;
+	tmp &= ~(1 << L1C_CACHEABLE_POS);
+	tmp &= ~L1C_WAY_DIS_MSK;
+	tmp &= ~(1 << L1C_CNT_EN_POS);
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+
+	tmp &= ~(1 << L1C_INVALID_EN_POS);
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+	__asm__ volatile (".rept 4 ; nop ; .endr");
+
+	tmp |= (1 << L1C_INVALID_EN_POS);
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+	__asm__ volatile (".rept 4 ; nop ; .endr");
+
+	while ((*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET))
+		& L1C_INVALID_DONE_MSK) == 0 && timeout > 0) {
+		clock_bflb_settle();
+		clock_bflb_settle();
+		timeout--;
+	}
+
+	if (!timeout) {
+		return -EIO;
+	}
+
+#ifdef SOC_SERIES_BL70X
+	tmp &= ~(1 << L1C_FLUSH_EN_POS);
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+	__asm__ volatile (".rept 4 ; nop ; .endr");
+	tmp |= (1 << L1C_FLUSH_EN_POS);
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+	__asm__ volatile (".rept 4 ; nop ; .endr");
+
+	while ((*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET))
+		& L1C_FLUSH_DONE_MSK) == 0 && timeout > 0) {
+		clock_bflb_settle();
+		clock_bflb_settle();
+		timeout--;
+	}
+
+	if (!timeout) {
+		return -EIO;
+	}
+
+	tmp &= ~(1 << L1C_FLUSH_EN_POS);
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+#endif
+
+	tmp = *((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET));
+	tmp |= 1 << L1C_BYPASS_POS;
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+
+	tmp &= ~(1 << L1C_BYPASS_POS);
+	tmp |= 1 << L1C_CNT_EN_POS;
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+
+	tmp &= ~L1C_WAY_DIS_MSK;
+	tmp |= cache_cfg.ways_mask << L1C_WAY_DIS_POS;
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
+
+	tmp |= 1 << L1C_CACHEABLE_POS;
+	*((volatile uint32_t *)(L1C_BASE + L1C_CONFIG_OFFSET)) = tmp;
```

Comment:
That looks like all matching  what the SDK does by looking at BL702, except this part:

```c
    /* If way disable is 0x0f, cacheable can't be set */
    if (finWayDisable != 0x0f) {
        tmpVal = BL_SET_REG_BIT(tmpVal, L1C_CACHEABLE);
    } else {
        tmpVal = BL_CLR_REG_BIT(tmpVal, L1C_CACHEABLE);
    }
```
https://github.com/bouffalolab/bouffalo_sdk/blob/811e8c1f1cd62695cedd47367f0bf2c624707bc6/drivers/soc/bl702/std/src/bl702_l1c.c#L178-L187

I assume this is a no-op if `L1C_CACHEABLE` is written anyway?